File: backend/mcp-servers/k8s/strata_mcp_k8s/crypto.py
"""AES-256-GCM decryption for Strata cluster credentials."""

from __future__ import annotations

import base64
import hashlib
import os

from cryptography.hazmat.primitives.ciphers.aead import AESGCM

DEFAULT_ENCRYPTION_SECRET = "strata-dev-insecure-master-key-change-me"


def derive_key(secret: str) -> bytes:
    """Derive a 32-byte AES-256 key from a secret using SHA-256."""
    return hashlib.sha256(secret.encode("utf-8")).digest()


def decrypt_kubeconfig(encrypted_b64: str, secret: str | None = None) -> str:
    """Decrypt a base64-encoded AES-256-GCM ciphertext containing a kubeconfig.

    Expects the first 12 bytes to be the nonce, followed by the ciphertext
    and 16-byte authentication tag (matching Go's crypto.Encrypt format).
    """
    if not secret:
        secret = os.environ.get("ENCRYPTION_SECRET", DEFAULT_ENCRYPTION_SECRET)
    key = derive_key(secret)
    raw = base64.b64decode(encrypted_b64)
    if len(raw) < 12:
        raise ValueError("encrypted payload too short to contain nonce")
    nonce = raw[:12]
    ciphertext = raw[12:]
    aesgcm = AESGCM(key)
    plaintext = aesgcm.decrypt(nonce, ciphertext, None)
    return plaintext.decode("utf-8")

File: backend/services/shared/pkg/crypto/crypto.go
// Package crypto provides AES-256-GCM authenticated encryption and decryption
// for Kubernetes cluster credentials stored in Strata.
package crypto

import (
	"crypto/aes"
	"crypto/cipher"
	"crypto/rand"
	"crypto/sha256"
	"encoding/base64"
	"errors"
	"fmt"
	"io"
)

// DeriveKey derives a 32-byte AES-256 key from a secret string using SHA-256.
func DeriveKey(secret string) []byte {
	h := sha256.Sum256([]byte(secret))
	return h[:]
}

// Encrypt encrypts plaintext using AES-256-GCM and returns a base64-encoded
// string containing the 12-byte nonce followed by ciphertext and authentication tag.
func Encrypt(key []byte, plaintext []byte) (string, error) {
	if len(key) != 32 {
		return "", errors.New("crypto: key must be 32 bytes for AES-256")
	}
	block, err := aes.NewCipher(key)
	if err != nil {
		return "", fmt.Errorf("crypto: new cipher: %w", err)
	}
	gcm, err := cipher.NewGCM(block)
	if err != nil {
		return "", fmt.Errorf("crypto: new gcm: %w", err)
	}
	nonce := make([]byte, gcm.NonceSize())
	if _, err := io.ReadFull(rand.Reader, nonce); err != nil {
		return "", fmt.Errorf("crypto: generate nonce: %w", err)
	}
	// gcm.Seal appends ciphertext + tag to nonce buffer
	sealed := gcm.Seal(nonce, nonce, plaintext, nil)
	return base64.StdEncoding.EncodeToString(sealed), nil
}

// Decrypt decodes a base64 string produced by Encrypt and decrypts the
// AES-256-GCM ciphertext using the provided key.
func Decrypt(key []byte, encoded string) ([]byte, error) {
	if len(key) != 32 {
		return nil, errors.New("crypto: key must be 32 bytes for AES-256")
	}
	raw, err := base64.StdEncoding.DecodeString(encoded)
	if err != nil {
		return nil, fmt.Errorf("crypto: decode base64: %w", err)
	}
	block, err := aes.NewCipher(key)
	if err != nil {
		return nil, fmt.Errorf("crypto: new cipher: %w", err)
	}
	gcm, err := cipher.NewGCM(block)
	if err != nil {
		return nil, fmt.Errorf("crypto: new gcm: %w", err)
	}
	nonceSize := gcm.NonceSize()
	if len(raw) < nonceSize {
		return nil, errors.New("crypto: ciphertext too short")
	}
	nonce, ciphertext := raw[:nonceSize], raw[nonceSize:]
	plaintext, err := gcm.Open(nil, nonce, ciphertext, nil)
	if err != nil {
		return nil, fmt.Errorf("crypto: decrypt: %w", err)
	}
	return plaintext, nil
}
